File: stack_Queue copy/circular_queue.c
#include <stdio.h>
#include <stdlib.h>
#define MAX 6

int queue[MAX];
int front = -1, rear = -1;

void insert(int key)
    {
    if ((front == 0 && rear == MAX - 1) || (front == rear + 1)) {
        puts("[Error]_Queue_Overflow__");
        exit(1);
        }
    else if ((rear == -1) || (front == -1)) {
        rear = front = 0;
        queue[rear] = key;
        }
    else if (rear == MAX - 1) {
        rear = 0;
        }
    else {
        queue[++rear] = key;
        }
    }

void delete()
    {
    if (front < 0) {
        puts("[Error]_Queue_Underflow__");
        exit(1);
        }
    else if (front == rear) {
        queue[front] = '\0';
        front = rear = -1;
        }
    else if (front == MAX - 1) {
        queue[front] = '\0';
        front = 0;
        }
    else {
        queue[front++] = '\0';
        }
    }

int main()
    {
    for (int i = 0;i < MAX ;i++) {
        insert((i + 1) * 10);
        }
    // printf("Front [%d] | Rear [%d]\n", front, rear);
        delete();
        delete();
    // printf("Front [%d] | Rear [%d]\n", front, rear);
    for (int i = 0;i < MAX;i++) {
        printf("%d\t", queue[i]);
        }
    }

File: go reimplementation/stack_queue/circular_queue.go
// circular queue
package main

import (
	"data_struct/color"
	"fmt"
)

var queue [64]rune
var front, rear = -1, -1

func insert(char rune) {
	switch rear {
	case front - 1:
		fmt.Println("Queue Overflow")
		return
	case -1:
		front = 0
		rear = 0
		queue[rear] = char
	case 63:
		if front == 0 {
			fmt.Println("Queue Overflow")
			return
		}
		rear = 0
		queue[rear] = char
	default:
		rear++
		queue[rear] = char
	}
}

func delete() {
	if front == -1 {
		fmt.Println("[ Error ] Queue underflow")
		return
	}
	queue[front] = 0
	if front == rear {
		front = -1
		rear = -1
		return
	}
	front = (front + 1) % 64
}
func printqueue() {
	fmt.Printf("%s %sQueue %s\n :", color.Bold, color.FgBrightMagenta, color.Reset)
	fmt.Println(queue)
}

func main() {
	fmt.Printf("Circular queue\n No.of Insertions :\t ")
	input := 0
	fmt.Scanln(&input)
	for i := 0; i < input; i++ {
		insert(rune(i))
	}
	printqueue()
	for i := 0; i < 10; i++ {
		delete()
	}
	printqueue()
	for i := 0; i < 9; i++ {
		insert(rune(i))
	}
	printqueue()

}
